Source/Particles/Collision/BinaryCollision/PhotonPhoton/PhotonPhotonCollisionFunc.H

```diff
@@ -0,0 +1,243 @@
+        amrex::ParticleReal const  /*q1*/, amrex::ParticleReal const  /*q2*/,
+        amrex::ParticleReal const  /*m1*/, amrex::ParticleReal const  /*m2*/,
+        amrex::Real const  dt, amrex::Real const dV,
+        index_type const cell_start_pair, index_type* AMREX_RESTRICT p_mask,
+        index_type* AMREX_RESTRICT p_pair_indices_1, index_type* AMREX_RESTRICT p_pair_indices_2,
+        amrex::ParticleReal* AMREX_RESTRICT p_pair_reaction_weight,
+        amrex::RandomEngine const& engine) const
+        {
+            amrex::ParticleReal * const AMREX_RESTRICT w1 = soa_1.m_rdata[PIdx::w];
+            amrex::ParticleReal * const AMREX_RESTRICT u1x = soa_1.m_rdata[PIdx::ux];
+            amrex::ParticleReal * const AMREX_RESTRICT u1y = soa_1.m_rdata[PIdx::uy];
+            amrex::ParticleReal * const AMREX_RESTRICT u1z = soa_1.m_rdata[PIdx::uz];
+
+            amrex::ParticleReal * const AMREX_RESTRICT w2 = soa_2.m_rdata[PIdx::w];
+            amrex::ParticleReal * const AMREX_RESTRICT u2x = soa_2.m_rdata[PIdx::ux];
+            amrex::ParticleReal * const AMREX_RESTRICT u2y = soa_2.m_rdata[PIdx::uy];
+            amrex::ParticleReal * const AMREX_RESTRICT u2z = soa_2.m_rdata[PIdx::uz];
+
+            // Number of macroparticles of each species
+            const int NI1 = I1e - I1s;
+            const int NI2 = I2e - I2s;
+            const int max_N = amrex::max(NI1,NI2);
+
+            int i1 = I1s;
+            int i2 = I2s;
+            int pair_index = cell_start_pair;
+
+            // Because the number of particles of each species is not always equal (NI1 != NI2
+            // in general), some macroparticles will be paired with multiple macroparticles of the
+            // other species and we need to decrease their weight accordingly.
+            // c1 corresponds to the minimum number of times a particle of species 1 will be paired
+            // with a particle of species 2. Same for c2.
+            const int c1 = amrex::max(NI2/NI1,1);
+            const int c2 = amrex::max(NI1/NI2,1);
+
+            // multiplier ratio to take into account unsampled pairs
+            int multiplier_ratio;
+            if (m_isSameSpecies)
+            {
+                multiplier_ratio = 2*max_N - 1;
+            }
+            else
+            {
+                multiplier_ratio = max_N;
+            }
+
+#if (defined WARPX_DIM_RZ)
+            amrex::ParticleReal * const AMREX_RESTRICT theta1 = soa_1.m_rdata[PIdx::theta];
+            amrex::ParticleReal * const AMREX_RESTRICT theta2 = soa_2.m_rdata[PIdx::theta];
+#endif
+
+            for (int k = 0; k < max_N; ++k)
+            {
+                // c1k : how many times the current particle of species 1 is paired with a particle
+                // of species 2. Same for c2k.
+                const int c1k = (k%NI1 < max_N%NI1) ? c1 + 1: c1;
+                const int c2k = (k%NI2 < max_N%NI2) ? c2 + 1: c2;
```

Comment:
We could also store the numbers `max_N%NI1` and `max_N%NI2` in integer variables defined outside the loop over `k`, to avoid recomputing the `%` operation at each loop iteration. I guess it's a small optimization, but why not.